x/upgrade/module.go

```diff
@@ -199,8 +205,18 @@ func ProvideModule(in UpgradeInputs) UpgradeOutputs {
 
 	// set the governance module account as the authority for conducting upgrades
 	k := keeper.NewKeeper(skipUpgradeHeights, in.Key, in.Cdc, homePath, nil, authority.String())
+	baseappOpt := func(app *baseapp.BaseApp) {
+		k.SetVersionSetter(app)
+	}
 	m := NewAppModule(k)
 	gh := govv1beta1.HandlerRoute{RouteKey: types.RouterKey, Handler: NewSoftwareUpgradeProposalHandler(k)}
 
-	return UpgradeOutputs{UpgradeKeeper: k, Module: m, GovHandler: gh}
+	return UpgradeOutputs{UpgradeKeeper: k, Module: m, GovHandler: gh, BaseAppOption: baseappOpt}
+}
+
+var versionMap *module.VersionMap = nil
+
+func PopulateVersionMap(modules map[string]appmodule.AppModule) {
```

Comment:
It seems like will miss some modules if they are added manually. But maybe that's okay?

Reply:
Updated, as discussed. When chains register manually module in app_v2, they can set manually the InitChainer and the upgrade module init genesis can deduplicate it.